### ISSUE TITLE
[Godot 4.0 Beta 16] Custom resource export vars do not allow for inheritance

### ISSUE THREAD

**Issue by ladybunne** [godotengine/godot#72219]
### Godot version

Godot 4.0 Beta 16

### System information

Win10

### Issue description

Hello! This is my first bug submission, apologies if it's a bit rough.

I noticed a change in behaviour regarding export vars of a custom resource type. In Beta 14 and prior, you're able to specify an export var of a _parent_ custom resource type, then give it a value of a _child_ resource type that inherits from the parent type. (Basic inheritance, pretty much). I've verified this on Beta 10.

As of Beta 15, which introduced a change to how custom resource loading works, and added the ability to quick load custom resources, export vars of custom resources seem to no longer respect inheritance. I've tested this on Beta 16, but the change itself was introduced in Beta 15, so it should be the same behaviour.

I've attached a minimal reproduction project that demonstrates the new (bugged) behaviour.

Please let me know if you need any more info, or if I've made a mistake somewhere! Thanks for all the work on the project!

### Steps to reproduce

1. Create a custom resource that extends from Resource ("FirstResource").
2. Create a second resource that extends from the first one ("SecondResource").
3. Create an instance of SecondResource.
4. Create an export var of type FirstResource on a node.
5. Instantiate that node in a scene.
6. Restart Godot, since it's fussy about newly created resource types.
7. Try and load your instance of SecondResource onto the node's FirstResource property.
8. Observe the type error (e.g. _The selected resource (SecondResource) does not match any type expected for this property (FirstResource)._).

### Minimal reproduction project

[CustomResourceInheritanceDemo.zip](https://github.com/godotengine/godot/files/10524227/CustomResourceInheritanceDemo.zip)

**Comment by bitsawer** [comment]
Looks like a regression, the editor resource handling received some improvents during that time period. Maybe related to https://github.com/godotengine/godot/pull/71683 or https://github.com/godotengine/godot/pull/71687?

**Comment by dalexeev** [comment]
Probably duplicate of #66987.

**Comment by ababen1** [comment]
having this problem as well

**Comment by ababen1** [comment]
i restartted the engine and now i can load custom resources with inheritance as normal, but i cant quick load them. when the inheritance didn't work, however, quick loading them did work. it seems like the two can't get along together, and i'm not sure what makes the engine "pick" one over the other

**Comment by quinnyo** [comment]
tl;dr on a newer build, forcing the script class cache to be regenerated fixes the problem (e.g. by removing the `.godot` dir)

I ran into this as well, first using a custom build from master shortly after beta 16. I just built a new version (from 935a6ef46b7c6af4239b6adce41b8198f051b57a) but the issue was still present.

Looking at the recent commits, the `global_script_class_cache.cfg` file and the error message, it's clear that the deeper inherited classes were being treated as if they inherited Resource directly. While the incorrect cache file is still in your project, the problem will persist. For fun I edited the class cache file manually to set the base class correctly, which fixed the problem.

The class cache does get generated correctly now (at least on the latest version) -- so removing the .godot directory (or maybe just the cache file) and opening the project should fix the problem.

**Comment by ababen1** [comment]
I deleted .godot and Im still having this problem on linux.. it works fine on windows though

**Comment by akien-mga** [comment]
This seems to be fixed in 4.0 RC 2.

There are still some caveats when the global script class cache is out of sync and having to restart the editor (see #72989), but the MRP in this issue worked fine out of the box.

If the `.godot` folder is deleted, then the project needs to be opened once in the editor, then closed, then reopened, and it works fine.

![image](https://user-images.githubusercontent.com/4701338/219646866-5934aef9-25d9-49bb-b034-9484816dc8ac.png)

**Comment by max99x** [comment]
I'm seeing this bug in 4.5, but it only occurred when the game was launched (from the editor), not in the editor itself. Deleting `global_script_class_cache.cfg` from `.godot` fixed it, even though the regenerated `global_script_class_cache.cfg` is identical.

**Comment by AThousandShips** [comment]
Please open a new bug report and give clear steps for reproducing the issue and an MRP if possible
